### PR TITLE
sql: don't stop monitors when we're panicking

### PR DESCRIPTION
**Pull request by andreimatei**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10481)

### CODE REVIEW

File: pkg/sql/planner.go [hunk 1]
```diff
@@ -235,6 +235,12 @@ func makeInternalPlanner(
 }
 
 func finishInternalPlanner(p *planner) {
+	// If we're panicking, don't bother stopping the MemoryMonitors. We're likely
+	// in a state where not all memory has been properly release to the monitors,
```
Comment (tamird) [inline pkg/sql/planner.go]: s/release/released/

File: pkg/sql/planner.go [hunk 2]
```diff
@@ -235,6 +235,12 @@ func makeInternalPlanner(
 }
 
 func finishInternalPlanner(p *planner) {
+	// If we're panicking, don't bother stopping the MemoryMonitors. We're likely
```
Comment (tamird) [inline pkg/sql/planner.go]: this is "not bothering" to do a bunch of things, not just stopping the memory monitors.

File: pkg/sql/planner.go [hunk 3]
```diff
@@ -235,6 +235,12 @@ func makeInternalPlanner(
...
+	// in a state where not all memory has been properly release to the monitors,
+	// so Stop() would panic again and add confusion to stack traces.
```
Comment (tamird) [inline pkg/sql/planner.go]: it's generally not a good idea to refer to specific bindings or names like `MemoryMonitors` and `Stop`; if these names change, the comment will have rotted, and the names' presence does not add much clarity in a majority of cases.

**Comment by knz** [comment]
Can you explain in the commit message what has motivated this change. Thanks.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-:-KVpuG-Y_ykBYPFTP0IR:birfkmz)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/planner.go, line 238 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-KVpsT2N9-SQ60xh1pNe-r1-238:-KVuDvnUIHN2i-l2fgOb:bsed0wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1333deab2f4e3b063894cbcbdbf428c0a38c0eb0/pkg/sql/planner.go#L238)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is "not bothering" to do a bunch of things, not just stopping the memory monitors.
</details>
not sure what you mean. The only thing in this function is stopping monitors.

---

*[pkg/sql/planner.go, line 239 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-KVpsT2N9-SQ60xh1pNe-r1-239:-KVuDtm40_qba2Hzk8ml:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1333deab2f4e3b063894cbcbdbf428c0a38c0eb0/pkg/sql/planner.go#L239)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/release/released/
</details>
Done.

---

*[pkg/sql/planner.go, line 240 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-KVpsT2N9-SQ60xh1pNe-r1-240:-KVuE45rRiWLZiv5gdcJ:b4lpxnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1333deab2f4e3b063894cbcbdbf428c0a38c0eb0/pkg/sql/planner.go#L240)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it's generally not a good idea to refer to specific bindings or names like `MemoryMonitors` and `Stop`; if these names change, the comment will have rotted, and the names' presence does not add much clarity in a majority of cases.
</details>
I recognize the desire to not be overly tied to name and specifics in general, but I think here it's fine; the phrase is worse without referring to Stop by name. Even if the name changes, the comment would still be fine (except in case there's no longer an opportunity to panic here, in which case the fact that the comment has visibly rotted would be a good thing).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481)*

**Comment by andreimatei** [comment]
expended the commit msg

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-:-KVuEyw5-aftYXlSBqvb:b-jb967v)*

**Comment by knz** [comment]
LGTM

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/planner.go, line 238 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-KVpsT2N9-SQ60xh1pNe-r1-238:-KVvUP4O3BGAJ8vVwqlO:b-jzq1uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1333deab2f4e3b063894cbcbdbf428c0a38c0eb0/pkg/sql/planner.go#L238)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> not sure what you mean. The only thing in this function is stopping monitors.
</details>
I mean that this function is called `finishInternalPlanner`, not `stopMemoryMonitors`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481)*

**Comment by knz** [comment]
Welp, I rescind my LGTM. Unfortunately, if you avoid stopping these monitors, the upper level monitors will complain that *they* see a memory leak.

What's really needed is a new+separate `StopOnPanic()` method which releases the allocated budget if there's any left.

**Comment by andreimatei** [comment]
Well somehow I didn't see the upper monitors complaining when I tested this. Maybe because they get stopped by a stopper that already skips doing that if it's recovering? I'll check.

**Comment by andreimatei** [comment]
@knz there's no "upper level monitors" here, are there? The memory comes seems to come from ether - see `makeInternalPlanner`. Right?

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/planner.go, line 238 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-KVpsT2N9-SQ60xh1pNe-r1-238:-KjOfGkPIJiH-rzTQImr:byqtrf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1333deab2f4e3b063894cbcbdbf428c0a38c0eb0/pkg/sql/planner.go#L238)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I mean that this function is called `finishInternalPlanner`, not `stopMemoryMonitors`.
</blockquote></details>

I've considered changing the name of the function, but the current one seems better to me. I've expended the comment to explain that we don't want to do the lines of code below on panic, not the whole function. Even though the lines below _are_ the whole function :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481#-:-KjOf6M70tfhxFrFGStl:b3sgvww)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10481)*

**Comment by knz** [comment]
This has been superseded by  #15358.

**Pull request closed without merging**